File: Java Playground/04_LinkedList.java
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Set;

/**
 * Linked List Problems - Complete Guide for Interview Preparation.
 *
 * <p>Core Concepts:
 * <ul>
 *   <li>Always check head == null and head.next == null edge cases.</li>
 *   <li>Dummy head node pattern eliminates special-casing of head modifications.</li>
 *   <li>Fast/slow pointer: fast moves 2 steps, slow moves 1 step (cycle, middle).</li>
 *   <li>Draw pointer diagrams mentally before coding in-place reversals.</li>
 *   <li>Time Complexity: O(n) traversal, O(n^2) nested loops. Space: O(1) for
 *       in-place, O(n) for recursion.</li>
 * </ul>
 *
 * <p>Tricky Parts:
 * <ul>
 *   <li>In-place reversal: maintain prev, curr, next pointers carefully.</li>
 *   <li>Cycle detection: Floyd's algorithm uses slow/fast pointers.</li>
 *   <li>Random pointer copy: HashMap to map original nodes to copies.</li>
 *   <li>Palindrome: find middle, reverse second half, then compare.</li>
 * </ul>
 *
 * @author Google Senior SWE Interview Playground
 */
public class LinkedList {

  /**
   * Definition for singly-linked list node.
   */
  public static class ListNode {
    int val;
    ListNode next;

    /**
     * Constructor with value only.
     *
     * @param val node value
     */
    public ListNode(int val) {
      this.val = val;
    }

    /**
     * Constructor with value and next pointer.
     *
     * @param val node value
     * @param next next node
     */
    public ListNode(int val, ListNode next) {
      this.val = val;
      this.next = next;
    }
  }

  /**
   * Definition for linked list node with random pointer (LC 138).
   */
  public static class RandomNode {
    int val;
    RandomNode next;
    RandomNode random;

    /**
     * Constructor for RandomNode.
     *
     * @param val node value
     */
    public RandomNode(int val) {
      this.val = val;
      this.next = null;
      this.random = null;
    }
  }

  /**
   * Detect if linked list has a cycle using Floyd's algorithm.
   *
   * <p>LeetCode 141 - Linked List Cycle (Easy)
   *
   * <p>Time Complexity: O(n), Space Complexity: O(1)
   *
   * @param head head of linked list
   * @return true if cycle exists, false otherwise
   * @throws IllegalArgumentException if head is null
   */
  public static boolean hasCycle(ListNode head) {
    if (head == null || head.next == null) {
      return false;
    }

    ListNode slow = head;
    ListNode fast = head;

    while (fast != null && fast.next != null) {
      slow = slow.next;
      fast = fast.next.next;

      if (slow == fast) {
        return true;
      }
    }

    return false;
  }

  /**
   * Detect cycle and return node where cycle begins.
   *
   * <p>LeetCode 142 - Linked List Cycle II (Medium)
   *
   * <p>Time Complexity: O(n), Space Complexity: O(1)
   *
   * <p>Pattern: Find meeting point with slow/fast pointers. Then reset one pointer to
   * head and move both one step at a time until they meet at cycle start.
   *
   * @param head head of linked list
   * @return node where cycle begins, or null if no cycle
   */
  public static ListNode detectCycle(ListNode head) {
    if (head == null || head.next == null) {
      return null;
    }

    ListNode slow = head;
    ListNode fast = head;

    while (fast != null && fast.next != null) {
      slow = slow.next;
      fast = fast.next.next;

      if (slow == fast) {
        ListNode ptr1 = head;
        ListNode ptr2 = slow;

        while (ptr1 != ptr2) {
          ptr1 = ptr1.next;
          ptr2 = ptr2.next;
        }

        return ptr1;
      }
    }

    return null;
  }

  /**
   * Find middle of linked list using slow/fast pointers.
   *
   * <p>Time Complexity: O(n), Space Complexity: O(1)
   *
   * @param head head of linked list
   * @return middle node (or first of two middle nodes if even length)
   */
  public static ListNode findMiddle(ListNode head) {
    if (head == null || head.next == null) {
      return head;
    }

    ListNode slow = head;
    ListNode fast = head;

    while (fast != null && fast.next != null) {
      slow = slow.next;
      fast = fast.next.next;
    }

    return slow;
  }

  /**
   * Remove Nth node from end of list.
   *
   * <p>LeetCode 19 - Remove Nth Node From End of List (Medium)
   *
   * <p>Time Complexity: O(n), Space Complexity: O(1)
   *
   * <p>Pattern: Use dummy head. Two-pointer technique with n-step gap.
   *
   * @param head head of linked list
   * @param n position from end to remove (1-indexed)
   * @return new head after removal
   * @throws IllegalArgumentException if head is null or n is invalid
   */
  public static ListNode removeNthFromEnd(ListNode head, int n) {
    if (head == null || n < 1) {
      throw new IllegalArgumentException("Invalid input");
    }

    ListNode dummy = new ListNode(0, head);
    ListNode first = dummy;
    ListNode second = dummy;

    for (int i = 0; i <= n; i++) {
      if (first == null) {
        throw new IllegalArgumentException("n is greater than list length");
      }
      first = first.next;
    }

    while (first != null) {
      first = first.next;
      second = second.next;
    }

    second.next = second.next.next;
    return dummy.next;
  }

  /**
   * Reverse entire linked list.
   *
   * <p>LeetCode 206 - Reverse Linked List (Easy)
   *
   * <p>Time Complexity: O(n), Space Complexity: O(1)
   *
   * @param head head of linked list
   * @return new head of reversed list
   */
  public static ListNode reverseList(ListNode head) {
    ListNode prev = null;
    ListNode curr = head;

    while (curr != null) {
      ListNode nextTemp = curr.next;
      curr.next = prev;
      prev = curr;
      curr = nextTemp;
    }

    return prev;
  }

  /**
   * Reverse linked list between positions left and right.
   *
   * <p>LeetCode 92 - Reverse Linked List II (Medium)
   *
   * <p>Time Complexity: O(n), Space Complexity: O(1)
   *
   * @param head head of linked list
   * @param left start position (1-indexed)
   * @param right end position (1-indexed)
   * @return head of modified list
   * @throws IllegalArgumentException if left or right invalid
   */
  public static ListNode reverseBetween(ListNode head, int left, int right) {
    if (head == null || left < 1 || left > right) {
      throw new IllegalArgumentException("Invalid left or right position");
    }

    ListNode dummy = new ListNode(0, head);
    ListNode prevLeft = dummy;

    for (int i = 1; i < left; i++) {
      if (prevLeft.next == null) {
        throw new IllegalArgumentException("Left position exceeds list length");
      }
      prevLeft = prevLeft.next;
    }

    ListNode leftNode = prevLeft.next;
    ListNode prevCurr = leftNode;
    ListNode curr = leftNode.next;

    for (int i = 0; i < right - left; i++) {
      if (curr == null) {
        throw new IllegalArgumentException("Right position exceeds list length");
      }
      ListNode nextTemp = curr.next;
      curr.next = prevCurr;
      prevCurr = curr;
      curr = nextTemp;
    }

    leftNode.next = curr;
    prevLeft.next = prevCurr;

    return dummy.next;
  }

  /**
   * Reverse nodes of k group.
   *
   * <p>LeetCode 25 - Reverse Nodes in k-Group (Hard)
   *
   * <p>Time Complexity: O(n), Space Complexity: O(1)
   *
   * @param head head of linked list
   * @param k group size
   * @return head of modified list
   * @throws IllegalArgumentException if head is null or k invalid
   */
  public static ListNode reverseKGroup(ListNode head, int k) {
    if (head == null || k < 1) {
      throw new IllegalArgumentException("Invalid input");
    }

    ListNode dummy = new ListNode(0, head);
    ListNode prevGroup = dummy;

    while (true) {
      ListNode groupEnd = prevGroup;
      for (int i = 0; i < k; i++) {
        groupEnd = groupEnd.next;
        if (groupEnd == null) {
          return dummy.next;
        }
      }

      ListNode nextGroup = groupEnd.next;
      ListNode curr = prevGroup.next;
      ListNode prev = nextGroup;

      for (int i = 0; i < k; i++) {
        ListNode nextTemp = curr.next;
        curr.next = prev;
        prev = curr;
        curr = nextTemp;
      }

      ListNode temp = prevGroup.next;
      prevGroup.next = groupEnd;
      prevGroup = temp;
    }
  }

  /**
   * Merge two sorted linked lists.
   *
   * <p>LeetCode 21 - Merge Two Sorted Lists (Easy)
   *
   * <p>Time Complexity: O(n + m), Space Complexity: O(1)
   *
   * @param list1 first sorted list
   * @param list2 second sorted list
   * @return merged sorted list
   */
  public static ListNode mergeTwoLists(ListNode list1, ListNode list2) {
    ListNode dummy = new ListNode(0);
    ListNode curr = dummy;

    while (list1 != null && list2 != null) {
      if (list1.val <= list2.val) {
        curr.next = list1;
        list1 = list1.next;
      } else {
        curr.next = list2;
        list2 = list2.next;
      }
      curr = curr.next;
    }

    if (list1 != null) {
      curr.next = list1;
    } else {
      curr.next = list2;
    }

    return dummy.next;
  }

  /**
   * Merge K sorted linked lists.
   *
   * <p>LeetCode 23 - Merge k Sorted Lists (Hard)
   *
   * <p>Time Complexity: O(n log k) where k is number of lists
   *
   * <p>Space Complexity: O(k) for priority queue
   *
   * @param lists array of sorted lists
   * @return merged sorted list
   * @throws IllegalArgumentException if lists is null
   */
  public static ListNode mergeKLists(ListNode[] lists) {
    if (lists == null || lists.length == 0) {
      throw new IllegalArgumentException("Input lists cannot be null or empty");
    }

    PriorityQueue<ListNode> minHeap = new PriorityQueue<>((a, b) ->
        Integer.compare(a.val, b.val));

    for (ListNode list : lists) {
      if (list != null) {
        minHeap.offer(list);
      }
    }

    ListNode dummy = new ListNode(0);
    ListNode curr = dummy;

    while (!minHeap.isEmpty()) {
      ListNode node = minHeap.poll();
      curr.next = node;
      curr = curr.next;

      if (node.next != null) {
        minHeap.offer(node.next);
      }
    }

    return dummy.next;
  }

  /**
   * Reorder list: L0 -> Ln -> L1 -> Ln-1 -> L2 -> Ln-2.
   *
   * <p>LeetCode 143 - Reorder List (Medium)
   *
   * <p>Time Complexity: O(n), Space Complexity: O(1)
   *
   * <p>Pattern: Find middle, reverse second half, merge two halves.
   *
   * @param head head of linked list
   * @throws IllegalArgumentException if head is null
   */
  public static void reorderList(ListNode head) {
    if (head == null || head.next == null) {
      return;
    }

    ListNode middle = findMiddle(head);
    ListNode secondHalf = middle.next;
    middle.next = null;

    secondHalf = reverseList(secondHalf);

    ListNode first = head;
    ListNode second = secondHalf;

    while (second != null) {
      ListNode firstNext = first.next;
      ListNode secondNext = second.next;

      first.next = second;
      second.next = firstNext;

      first = firstNext;
      second = secondNext;
    }
  }

  /**
   * Odd-Even Linked List: separate odd and even positioned nodes.
   *
   * <p>LeetCode 328 - Odd Even Linked List (Medium)
   *
   * <p>Time Complexity: O(n), Space Complexity: O(1)
   *
   * @param head head of linked list
   * @return head of modified list (odd nodes first, then even nodes)
   */
  public static ListNode oddEvenList(ListNode head) {
    if (head == null || head.next == null) {
      return head;
    }

    ListNode oddHead = head;
    ListNode evenHead = head.next;
    ListNode oddTail = oddHead;
    ListNode evenTail = evenHead;

    ListNode curr = head.next.next;
    boolean isOdd = true;

    while (curr != null) {
      if (isOdd) {
        oddTail.next = curr;
        oddTail = oddTail.next;
      } else {
        evenTail.next = curr;
        evenTail = evenTail.next;
      }

      curr = curr.next;
      isOdd = !isOdd;
    }

    oddTail.next = evenHead;
    evenTail.next = null;

    return oddHead;
  }

  /**
   * Copy list with random pointer.
   *
   * <p>LeetCode 138 - Copy List with Random Pointer (Medium)
   *
   * <p>Time Complexity: O(n), Space Complexity: O(n)
   *
   * <p>Pattern: First pass create copies and store in map. Second pass set next/random
   * pointers using map.
   *
   * @param head head of list with random pointers
   * @return deep copy of list
   */
  public static RandomNode copyRandomList(RandomNode head) {
    if (head == null) {
      return null;
    }

    Map<RandomNode, RandomNode> nodeMap = new HashMap<>();
    RandomNode curr = head;

    while (curr != null) {
      nodeMap.put(curr, new RandomNode(curr.val));
      curr = curr.next;
    }

    curr = head;
    while (curr != null) {
      RandomNode copy = nodeMap.get(curr);
      copy.next = nodeMap.get(curr.next);
      copy.random = nodeMap.get(curr.random);
      curr = curr.next;
    }

    return nodeMap.get(head);
  }

  /**
   * Check if linked list is a palindrome using slow/fast pointers.
   *
   * <p>LeetCode 234 - Palindrome Linked List (Easy)
   *
   * <p>Time Complexity: O(n), Space Complexity: O(1)
   *
   * <p>Pattern: Find middle, reverse second half, compare two halves.
   *
   * @param head head of linked list
   * @return true if list is palindrome
   */
  public static boolean isPalindrome(ListNode head) {
    if (head == null || head.next == null) {
      return true;
    }

    ListNode middle = findMiddle(head);
    ListNode secondHalf = reverseList(middle.next);

    ListNode first = head;
    ListNode second = secondHalf;

    while (second != null) {
      if (first.val != second.val) {
        return false;
      }
      first = first.next;
      second = second.next;
    }

    return true;
  }

  /**
   * Helper to print linked list for testing.
   *
   * @param head head of list
   * @return string representation
   */
  private static String printList(ListNode head) {
    StringBuilder sb = new StringBuilder();
    ListNode curr = head;
    while (curr != null) {
      sb.append(curr.val).append(" -> ");
      curr = curr.next;
    }
    sb.append("null");
    return sb.toString();
  }

  /**
   * Main method with comprehensive test cases.
   *
   * @param args command-line arguments (unused)
   */
  public static void main(String[] args) {
    System.out.println("=== Linked List Problems ===\n");

    // Test reverseList
    System.out.println("1. Reverse Linked List:");
    ListNode head1 = new ListNode(1, new ListNode(2, new ListNode(3,
        new ListNode(4))));
    System.out.println("  Input: " + printList(head1));
    ListNode reversed = reverseList(head1);
    System.out.println("  Output: " + printList(reversed) + "\n");

    // Test hasCycle
    System.out.println("2. Detect Cycle:");
    ListNode cycleNode = new ListNode(3);
    ListNode head2 = new ListNode(1, new ListNode(2, cycleNode));
    cycleNode.next = head2.next;
    boolean hasCycle = hasCycle(head2);
    System.out.println("  List with cycle: " + hasCycle + "\n");

    // Test findMiddle
    System.out.println("3. Find Middle:");
    ListNode head3 = new ListNode(1, new ListNode(2, new ListNode(3,
        new ListNode(4, new ListNode(5)))));
    ListNode middle = findMiddle(head3);
    System.out.println("  Input: " + printList(head3));
    System.out.println("  Middle value: " + middle.val + "\n");

    // Test mergeTwoLists
    System.out.println("4. Merge Two Sorted Lists:");
    ListNode list1 = new ListNode(1, new ListNode(2, new ListNode(4)));
    ListNode list2 = new ListNode(1, new ListNode(3, new ListNode(4)));
    ListNode merged = mergeTwoLists(list1, list2);
    System.out.println("  Output: " + printList(merged) + "\n");

    // Test reorderList
    System.out.println("5. Reorder List:");
    ListNode head5 = new ListNode(1, new ListNode(2, new ListNode(3,
        new ListNode(4))));
    System.out.println("  Input: " + printList(head5));
    reorderList(head5);
    System.out.println("  Output: " + printList(head5) + "\n");

    // Test isPalindrome
    System.out.println("6. Palindrome Check:");
    ListNode head6 = new ListNode(1, new ListNode(2, new ListNode(2,
        new ListNode(1))));
    System.out.println("  Input: " + printList(head6));
    boolean isPal = isPalindrome(head6);
    System.out.println("  Is Palindrome: " + isPal + "\n");

    // Test oddEvenList
    System.out.println("7. Odd Even Linked List:");
    ListNode head7 = new ListNode(1, new ListNode(2, new ListNode(3,
        new ListNode(4, new ListNode(5)))));
    System.out.println("  Input: " + printList(head7));
    ListNode oddEven = oddEvenList(head7);
    System.out.println("  Output: " + printList(oddEven) + "\n");
  }
}

File: C++ Playground/04_LinkedList.cpp
/// @file 04_LinkedList.cpp
/// @brief Linked Lists Playground — reversal, cycle detection, merge,
///        reorder, copy with random, palindrome.
///
/// LINKED LISTS GUIDE:
/// - Define ListNode struct at top with constructor (val, next, NOT val_, next_)
/// - Always check for nullptr before dereferencing next or random pointers
/// - Use dummy head node to simplify edge cases (ListNode dummy(0); ...)
/// - Fast-and-slow pointer: fast jumps 2, slow jumps 1, meet at cycle/middle
/// - Reversal: track prev, curr, next; prev follows curr; curr follows next
/// - Merge with heap: use priority_queue<ListNode*> with custom comparator
/// - Copy with random: interleave original and copy, then separate
/// - LeetCode struct names: ListNode uses val, next, RandomNode uses val,
///   next, random (NO trailing underscores)

#include <iostream>
#include <queue>
#include <unordered_map>
#include <vector>

namespace linkedlist_playground {

/// @brief Singly-linked list node (LeetCode API compatible).
struct ListNode {
  int val;
  ListNode* next;
  explicit ListNode(int x) : val(x), next(nullptr) {}
};

/// @brief Node with random pointer (LC 138, LeetCode API compatible).
struct RandomNode {
  int val;
  RandomNode* next;
  RandomNode* random;
  explicit RandomNode(int x) : val(x), next(nullptr), random(nullptr) {}
};

/// @brief Detect cycle in linked list using fast-slow pointers (LC 141 Easy).
/// @param head Head of linked list.
/// @return true if cycle exists, false otherwise.
///
/// Example: 1→2→3→4→2 (cycle) → true; 1→2→3 (no cycle) → false
/// Constraints: 0 ≤ n ≤ 10⁴; −10⁵ ≤ val ≤ 10⁵.
/// Time: O(n), Space: O(1)
/// Tricky: Fast pointer jumps 2 steps, slow jumps 1. If they meet, cycle.
///         If fast reaches nullptr, no cycle.
[[nodiscard]] bool HasCycle(ListNode* head) noexcept {
  ListNode* slow = head;
  ListNode* fast = head;
  while (fast && fast->next) {
    slow = slow->next;
    fast = fast->next->next;
    if (slow == fast) return true;
  }
  return false;
}

/// @brief Detect and return node where cycle begins (LC 142 Medium).
/// @param head Head of linked list.
/// @return Pointer to cycle node, or nullptr if no cycle.
///
/// Example: 1→2→3→4→2 (cycle at 2) → pointer to node with val=2
/// Constraints: 0 ≤ n ≤ 10⁴.
/// Time: O(n), Space: O(1)
/// Tricky: Meet point = fast and slow collide. Then move slow to head,
///         fast stays; move both 1 step at a time; meet at cycle start.
[[nodiscard]] ListNode* DetectCycle(ListNode* head) noexcept {
  ListNode* slow = head;
  ListNode* fast = head;

  while (fast && fast->next) {
    slow = slow->next;
    fast = fast->next->next;
    if (slow == fast) break;
  }

  if (!fast || !fast->next) return nullptr;

  slow = head;
  while (slow != fast) {
    slow = slow->next;
    fast = fast->next;
  }
  return slow;
}

/// @brief Find middle of linked list using fast-slow pointers.
/// @param head Head of linked list.
/// @return Middle node (for even length, return second middle).
///
/// Example: 1→2→3→4→5 → 3; 1→2→3→4 → 3
/// Constraints: 1 ≤ n ≤ 500.
/// Time: O(n), Space: O(1)
/// Tricky: Fast reaches end, slow at middle.
[[nodiscard]] ListNode* FindMiddle(ListNode* head) noexcept {
  ListNode* slow = head;
  ListNode* fast = head;

  while (fast && fast->next) {
    slow = slow->next;
    fast = fast->next->next;
  }
  return slow;
}

/// @brief Remove Nth node from end of list (LC 19 Medium).
/// @param head Head of linked list.
/// @param n Position from end (1-indexed).
/// @return Head of modified list.
///
/// Example: 1→2→3→4→5, n=2 → 1→2→3→5
/// Constraints: 1 ≤ n ≤ list length; may remove head.
/// Time: O(n), Space: O(1)
/// Tricky: Use dummy head; two pointers n steps apart; move both until
///         first reaches end; remove by skipping.
[[nodiscard]] ListNode* RemoveNthFromEnd(ListNode* head, int n) noexcept {
  ListNode dummy(0);
  dummy.next = head;
  ListNode* first = &dummy;
  ListNode* second = &dummy;

  for (int i = 0; i <= n; ++i) {
    if (!first) return head;
    first = first->next;
  }

  while (first) {
    first = first->next;
    second = second->next;
  }

  second->next = second->next->next;
  return dummy.next;
}

/// @brief Reverse linked list (LC 206 Easy).
/// @param head Head of linked list.
/// @return Head of reversed list.
///
/// Example: 1→2→3→4→5 → 5→4→3→2→1
/// Constraints: 0 ≤ n ≤ 5000.
/// Time: O(n), Space: O(1)
/// Tricky: Track prev, curr, next. Reverse link from curr→prev. Move all
///         forward.
[[nodiscard]] ListNode* ReverseList(ListNode* head) noexcept {
  ListNode* prev = nullptr;
  ListNode* curr = head;

  while (curr) {
    ListNode* next = curr->next;
    curr->next = prev;
    prev = curr;
    curr = next;
  }
  return prev;
}

/// @brief Reverse list between positions left and right (LC 92 Medium).
/// @param head Head of linked list.
/// @param left Start position (1-indexed).
/// @param right End position (1-indexed).
/// @return Head of modified list.
///
/// Example: 1→2→3→4→5, left=2, right=4 → 1→4→3→2→5
/// Constraints: 1 ≤ left ≤ right ≤ n.
/// Time: O(n), Space: O(1)
/// Tricky: Find node before left; reverse [left..right] using standard
///         reversal; connect back.
[[nodiscard]] ListNode* ReverseBetween(ListNode* head, int left,
                                        int right) noexcept {
  ListNode dummy(0);
  dummy.next = head;
  ListNode* prev_left = &dummy;

  for (int i = 0; i < left - 1; ++i) {
    prev_left = prev_left->next;
  }

  ListNode* curr = prev_left->next;
  for (int i = 0; i < right - left; ++i) {
    ListNode* next = curr->next;
    curr->next = next->next;
    next->next = prev_left->next;
    prev_left->next = next;
  }

  return dummy.next;
}

/// @brief Reverse k group: reverse every k elements (LC 25 Hard).
/// @param head Head of linked list.
/// @param k Group size.
/// @return Head of modified list.
///
/// Example: 1→2→3→4→5, k=2 → 2→1→4→3→5
/// Constraints: 1 ≤ n ≤ 5000; 1 ≤ k ≤ n.
/// Time: O(n), Space: O(1)
/// Tricky: Check if k nodes available; reverse them; reconnect groups.
[[nodiscard]] ListNode* ReverseKGroup(ListNode* head, int k) noexcept {
  ListNode dummy(0);
  dummy.next = head;
  ListNode* prev_group = &dummy;

  while (true) {
    ListNode* kth = prev_group;
    for (int i = 0; i < k; ++i) {
      if (!kth->next) return dummy.next;
      kth = kth->next;
    }

    ListNode* next_group = kth->next;
    ListNode* prev = next_group;
    ListNode* curr = prev_group->next;

    for (int i = 0; i < k; ++i) {
      ListNode* next = curr->next;
      curr->next = prev;
      prev = curr;
      curr = next;
    }

    ListNode* temp = prev_group->next;
    prev_group->next = kth;
    prev_group = temp;
  }
}

/// @brief Merge two sorted linked lists (LC 21 Easy).
/// @param list1 Head of first sorted list.
/// @param list2 Head of second sorted list.
/// @return Head of merged sorted list.
///
/// Example: 1→2→4, 1→3→4 → 1→1→2→3→4→4
/// Constraints: 0 ≤ n, m ≤ 50.
/// Time: O(n + m), Space: O(1)
/// Tricky: Use dummy head; compare nodes, append smaller; advance that
///         pointer.
[[nodiscard]] ListNode* MergeTwoLists(ListNode* list1,
                                       ListNode* list2) noexcept {
  ListNode dummy(0);
  ListNode* curr = &dummy;

  while (list1 && list2) {
    if (list1->val <= list2->val) {
      curr->next = list1;
      list1 = list1->next;
    } else {
      curr->next = list2;
      list2 = list2->next;
    }
    curr = curr->next;
  }

  curr->next = list1 ? list1 : list2;
  return dummy.next;
}

/// @brief Merge k sorted linked lists (LC 23 Hard).
/// @param lists Vector of k sorted list heads.
/// @return Head of merged sorted list.
///
/// Example: [[1,4,5],[1,3,4],[2,6]] → [1,1,2,1,3,4,4,5,6]
/// Constraints: 0 ≤ k ≤ 10⁴; 0 ≤ n_i ≤ 500.
/// Time: O(n log k) where n = total nodes, k = number of lists
/// Space: O(k) for heap
/// Tricky: Use min-heap to always get smallest node; pop, add its next.
[[nodiscard]] ListNode* MergeKLists(std::vector<ListNode*>& lists) {
  auto cmp = [](ListNode* a, ListNode* b) {
    return a->val > b->val;
  };
  std::priority_queue<ListNode*, std::vector<ListNode*>, decltype(cmp)>
      min_heap(cmp);

  for (ListNode* head : lists) {
    if (head) min_heap.push(head);
  }

  ListNode dummy(0);
  ListNode* curr = &dummy;

  while (!min_heap.empty()) {
    ListNode* node = min_heap.top();
    min_heap.pop();
    curr->next = node;
    curr = curr->next;
    if (node->next) {
      min_heap.push(node->next);
    }
  }

  return dummy.next;
}

/// @brief Reorder list: 1→2→3→4→5 becomes 1→5→2→4→3 (LC 143 Medium).
/// @param head Head of linked list (modified in-place).
///
/// Example: 1→2→3→4→5 → 1→5→2→4→3
/// Constraints: 1 ≤ n ≤ 5×10⁴; n ≥ 2.
/// Time: O(n), Space: O(1)
/// Tricky: Find middle; reverse second half; merge two halves.
void ReorderList(ListNode* head) noexcept {
  ListNode* slow = head;
  ListNode* fast = head;
  while (fast && fast->next) {
    slow = slow->next;
    fast = fast->next->next;
  }

  ListNode* prev = nullptr;
  ListNode* curr = slow;
  while (curr) {
    ListNode* next = curr->next;
    curr->next = prev;
    prev = curr;
    curr = next;
  }

  ListNode* first = head;
  ListNode* second = prev;
  while (second->next) {
    ListNode* next_first = first->next;
    ListNode* next_second = second->next;
    first->next = second;
    second->next = next_first;
    first = next_first;
    second = next_second;
  }
}

/// @brief Odd-even linked list: group odd-indexed, then even-indexed
///        (LC 328 Med).
/// @param head Head of linked list (modified in-place).
/// @return Head of reordered list.
///
/// Example: 1→2→3→4→5 → 1→3→5→2→4
/// Constraints: 0 ≤ n ≤ 10⁴.
/// Time: O(n), Space: O(1)
/// Tricky: Two pointers for odd/even chains; interleave.
[[nodiscard]] ListNode* OddEvenList(ListNode* head) noexcept {
  if (!head || !head->next) return head;

  ListNode* odd = head;
  ListNode* even = head->next;
  ListNode* even_head = even;

  while (even && even->next) {
    odd->next = odd->next->next;
    even->next = even->next->next;
    odd = odd->next;
    even = even->next;
  }

  odd->next = even_head;
  return head;
}

/// @brief Copy list with random pointer: deep copy (LC 138 Medium).
/// @param head Head of RandomNode list.
/// @return Head of deep copied list.
///
/// Example: 1(random→3)→2(random→1)→3(random→3) →
///          deep copy with same structure
/// Constraints: 0 ≤ n ≤ 100; next and random can point anywhere.
/// Time: O(n), Space: O(n) for copy
/// Tricky: Interleave: insert copy right after original in original list;
///         set random pointers on copy; then separate lists.
[[nodiscard]] RandomNode* CopyRandomList(RandomNode* head) {
  if (!head) return nullptr;

  for (RandomNode* curr = head; curr; curr = curr->next->next) {
    RandomNode* copy = new RandomNode(curr->val);
    copy->next = curr->next;
    curr->next = copy;
  }

  for (RandomNode* curr = head; curr; curr = curr->next->next) {
    RandomNode* copy = curr->next;
    if (curr->random) {
      copy->random = curr->random->next;
    }
  }

  RandomNode dummy(0);
  RandomNode* copy_prev = &dummy;
  for (RandomNode* curr = head; curr; curr = curr->next) {
    RandomNode* copy = curr->next;
    copy_prev->next = copy;
    copy_prev = copy;
    curr->next = curr->next->next;
  }

  return dummy.next;
}

/// @brief Check if linked list is palindrome (LC 234 Easy).
/// @param head Head of linked list.
/// @return true if list is palindrome, false otherwise.
///
/// Example: 1→2→2→1 → true; 1→2 → false
/// Constraints: 1 ≤ n ≤ 10⁵; −10⁵ ≤ val ≤ 10⁵.
/// Time: O(n), Space: O(1)
/// Tricky: Find middle; reverse second half in-place; compare.
[[nodiscard]] bool IsPalindrome(ListNode* head) noexcept {
  ListNode* slow = head;
  ListNode* fast = head;
  while (fast && fast->next) {
    slow = slow->next;
    fast = fast->next->next;
  }

  ListNode* prev = nullptr;
  ListNode* curr = slow;
  while (curr) {
    ListNode* next = curr->next;
    curr->next = prev;
    prev = curr;
    curr = next;
  }

  ListNode* left = head;
  ListNode* right = prev;
  while (right) {
    if (left->val != right->val) return false;
    left = left->next;
    right = right->next;
  }
  return true;
}

}  // namespace linkedlist_playground

int main() {
  std::cout << "=== LINKED LIST PLAYGROUND ===\n\n";

  linkedlist_playground::ListNode* head1 =
      new linkedlist_playground::ListNode(1);
  head1->next = new linkedlist_playground::ListNode(2);
  head1->next->next = new linkedlist_playground::ListNode(3);

  bool has_cycle = linkedlist_playground::HasCycle(head1);
  std::cout << "HasCycle(1→2→3): " << (has_cycle ? "true" : "false") << "\n";

  linkedlist_playground::ListNode* middle =
      linkedlist_playground::FindMiddle(head1);
  std::cout << "FindMiddle(1→2→3): val=" << middle->val << "\n";

  linkedlist_playground::ListNode* reversed =
      linkedlist_playground::ReverseList(head1);
  std::cout << "ReverseList(1→2→3): ";
  linkedlist_playground::ListNode* temp = reversed;
  while (temp) {
    std::cout << temp->val << "→";
    temp = temp->next;
  }
  std::cout << "\b\b  \n";

  linkedlist_playground::ListNode* list1 =
      new linkedlist_playground::ListNode(1);
  list1->next = new linkedlist_playground::ListNode(3);
  linkedlist_playground::ListNode* list2 =
      new linkedlist_playground::ListNode(2);
  list2->next = new linkedlist_playground::ListNode(4);
  linkedlist_playground::ListNode* merged =
      linkedlist_playground::MergeTwoLists(list1, list2);
  std::cout << "MergeTwoLists(1→3, 2→4): ";
  temp = merged;
  while (temp) {
    std::cout << temp->val << "→";
    temp = temp->next;
  }
  std::cout << "\b\b  \n";

  linkedlist_playground::ListNode* palin =
      new linkedlist_playground::ListNode(1);
  palin->next = new linkedlist_playground::ListNode(2);
  palin->next->next = new linkedlist_playground::ListNode(2);
  palin->next->next->next = new linkedlist_playground::ListNode(1);
  bool is_palin = linkedlist_playground::IsPalindrome(palin);
  std::cout << "IsPalindrome(1→2→2→1): " << (is_palin ? "true" : "false")
            << "\n";

  return 0;
}
